crates/ty_project/src/lib.rs

```diff
@@ -549,6 +569,7 @@ impl Project {
 enum ProjectFiles<'a> {
     OpenFiles(&'a FxHashSet<File>),
     Indexed(files::Indexed<'a>),
+    IndexedAndOpenVirtualFiles(files::Indexed<'a>, Option<&'a FxHashSet<File>>),
```

Comment:
I think I would simply combine those and eagerly compute the Virtual files in `ProjectFiles::new`.